### PR TITLE
chore: convert event list module to MVVM

### LINKED ISSUE #1056: Convert Event List to View Model
**Is it a ?**

- [ ] Bug Report

- [ ] Feature Request
    
- [x] Chore
    
### Chore

**Description**

Event list module should be converted to MVVM so View Pager can be implemented without any messy logic on the presenter side.

**Would you like to work on the issue?**

Yes

### LINKED ISSUE #1121: Save viewpager position in the Events list
### Bug Report    

**Actual Behaviour**

When we rotate the screen on the Events list screen, the ViewPager automatically moves to the `Live` section.

**Expected Behaviour**

It should stay in the same tab.

**Steps to reproduce it**

Open Events list and move to the `Past` tab. Now rotate the screen. It will move back to the `Live` tab.

**Would you like to work on the issue?**

No.

### PR DESCRIPTION
**Pull request by mishuvs**
Fixes #1056 

Checklist:

- [x] I have checked for PMD and check-style issues 
- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream development branch.

Changes: 
1. Converts events list module to MVVM
2. Adds test for EventsViewModel
2. Fixes #1121

### CODE REVIEW

File: app/src/test/java/org/fossasia/openevent/app/core/presenter/EventsPresenterTest.java [hunk 1]
```diff
@@ -1,204 +0,0 @@
-package org.fossasia.openevent.app.core.presenter;
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/openevent/app/core/presenter/EventsPresenterTest.java]: Add a test as well

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsViewModel.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        return progress;
+    }
+
+    public LiveData<Boolean> getSuccess() {
+        return success;
+    }
+
+    private Observable<Event> getEventSource(boolean forceReload) {
+        if (!forceReload && !events.getValue().isEmpty())
+            return Observable.fromIterable(events.getValue());
```
Comment (srv-twry) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsViewModel.java]: Once you've fetched the events list, what your code will do is as follows:
1. Get the list of events from the MutableLiveData.
2. Create an observable from it.
3. Then set it back to the MutableLiveData using `events.setValue(newEvents);`

Is it the optimal behaviour ?

This `getEventSource` was something that was needed in the presenter without the LiveData. We don't need it when using LiveData.

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsListAdapter.java [hunk 3]
```diff
@@ -28,26 +28,32 @@
...
     }
 
     public void categorizeEvents() {
 
         notifyDataSetChanged();
     }
 
+    public void setEvents(List<Event> events) {
+        this.events = events;
+        notifyDataSetChanged();
```
Comment (srv-twry) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsListAdapter.java]: This is not an optimal way to do it. Use `DiffUtil`. For more info, check the android architecture samples.

Comment (srv-twry) [reply app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsListAdapter.java]: https://medium.com/@iammert/using-diffutil-in-android-recyclerview-bdca8e4fbb00

This is good too.

**Comment by mishuvs** [comment]
@srv-twry Please check

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsDiffCallback.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package org.fossasia.openevent.app.core.event.list;
+
+import android.support.v7.util.DiffUtil;
+
+import org.fossasia.openevent.app.data.event.Event;
+
+import java.util.List;
+
+public class EventsDiffCallback extends DiffUtil.Callback {
```
Comment (srv-twry) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsDiffCallback.java]: Why a whole new class for this ?

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsListAdapter.java [hunk 5]
```diff
@@ -28,26 +29,38 @@
...
+    EventsListAdapter(List<Event> events, Bus bus, EventListFragment fragment) {
         this.events = events;
         this.bus = bus;
-        this.eventsPresenter = eventsPresenter;
+        this.fragment = fragment;
     }
 
-    public void categorizeEvents() {
-
+    public void setEvents(List<Event> events) {
```
Comment (srv-twry) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsListAdapter.java]: Can't quite figure out the use of this method. Where are you using it ? You can simply use `updateList` only.

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsDiffCallback.java (hunk 4 above)
Comment (srv-twry) [reply app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsDiffCallback.java]: Ignore my comment. It looks good in a separate class. :)

**Comment by mishuvs** [comment]
@srv-twry Please review

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsViewModel.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    public void sortBy(int criteria) {
+        if (criteria == SORTBYNAME)
+            Collections.sort(events.getValue(), (e1, e2) -> e1.getName().compareToIgnoreCase(e2.getName()));
+        else {
+            Collections.sort(events.getValue(), DateService::compareEventDates);
+        }
+    }
+
+    public void loadUserEvents(boolean forceReload) {
+
```
Comment (srv-twry) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsViewModel.java]: Unnecesary.

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsListAdapter.java [hunk 7]
```diff
@@ -28,24 +29,51 @@
...
-    public void categorizeEvents() {
+    public void updateList(List<Event> newEvents) {
+        if (events == null) {
+            events = newEvents;
+            notifyItemRangeInserted(0, newEvents.size());
+        } else {
+            DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
+                @Override
+                public int getOldListSize() {
+                    return events == null ? 0 : events.size();
```
Comment (srv-twry) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsListAdapter.java]: events will never be null in this case. You're already doing the null check in the `if` condition.

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsViewModel.java (hunk 6 above)
Comment (mishuvs) [reply app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsViewModel.java]: why?

Comment (srv-twry) [reply app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsViewModel.java]: I am referring to the blank line.

**Pull request merged**